core/src/test/java/kafka/log/remote/RemoteLogOffsetReaderTest.java

```diff
@@ -0,0 +1,196 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import scala.Option;
+import scala.util.Either;
+
+import static org.apache.kafka.common.record.FileRecords.TimestampAndOffset;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class RemoteLogOffsetReaderTest {
+
+    private final MockTime time = new MockTime();
+    private final TopicPartition topicPartition = new TopicPartition("test", 0);
+    private Path logDir;
+    private LeaderEpochFileCache cache;
+    private MockRemoteLogManager rlm;
+
+    @BeforeEach
+    void setUp() throws IOException {
+        logDir = Files.createTempDirectory("kafka-test");
+        LeaderEpochCheckpointFile checkpoint = new LeaderEpochCheckpointFile(TestUtils.tempFile(), new LogDirFailureChannel(1));
+        cache = new LeaderEpochFileCache(topicPartition, checkpoint, time.scheduler);
+        rlm = new MockRemoteLogManager(2, 10, logDir.toString());
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        rlm.close();
+        Utils.delete(logDir.toFile());
+    }
+
+    @Test
+    public void testReadRemoteLog() throws Exception {
+        AsyncOffsetReadFutureHolder<Either<Exception, Option<TimestampAndOffset>>> asyncOffsetReadFutureHolder =
+                rlm.asyncOffsetRead(topicPartition, time.milliseconds(), 0L, cache, Option::empty);
+        asyncOffsetReadFutureHolder.taskFuture().get(1, TimeUnit.SECONDS);
+        assertTrue(asyncOffsetReadFutureHolder.taskFuture().isDone());
+
+        Either<Exception, Option<TimestampAndOffset>> result = asyncOffsetReadFutureHolder.taskFuture().get();
+        assertFalse(result.isLeft());
+        assertTrue(result.isRight());
+        assertEquals(Option.apply(new TimestampAndOffset(100L, 90L, Optional.of(3))),
+                result.right().get());
+    }
+
+    @Test
+    public void testTaskQueueFullAndCancelTask() throws Exception {
+        rlm.pause();
+
+        List<AsyncOffsetReadFutureHolder<Either<Exception, Option<TimestampAndOffset>>>> holderList = new ArrayList<>();
+        // Task queue size is 10 and number of threads is 2, so it can accept at-most 12 items
+        for (int i = 0; i < 12; i++) {
+            holderList.add(rlm.asyncOffsetRead(topicPartition, time.milliseconds(), 0L, cache, Option::empty));
+        }
+        assertThrows(TimeoutException.class, () -> holderList.get(0).taskFuture().get(10, TimeUnit.MILLISECONDS));
+        assertEquals(0, holderList.stream().filter(h -> h.taskFuture().isDone()).count());
+
+        assertThrows(RejectedExecutionException.class, () ->
+                holderList.add(rlm.asyncOffsetRead(topicPartition, time.milliseconds(), 0L, cache, Option::empty)));
+
+        holderList.get(2).jobFuture().cancel(false);
```

Comment:
Why does the job cancel doesn't complete the task?

Reply:
The reason for maintaining 2 futures: `jobFuture` and `taskFuture` is to trigger the delayed operation completion (`delayedRemoteListOffsetsPurgatory#checkAndComplete(key)`) in the same remote-log-reader thread. When all the partitions results are received, then the delayedOperation gets completed.

There is no relationship between the `jobFuture` and `taskFuture`. We call `jobFuture#cancel` only when the delayedOperation gets expired. So, either completing the `taskFuture` or not won't have any effect as the `taskFuture` result won't be used.

Reply:
I see. Thanks.